test/MatrixFields/scalar_fieldmatrix.jl

```diff
@@ -0,0 +1,218 @@
+        Two_fields{
+            Two_fields{Float64, Float64},
+            Two_fields{Float64, Two_fields{Float64, Singleton{Float64}}},
+        },
+        Float64,
+    )
+    test_broadcasted_get_field_type(
+        @name(y.y.y),
+        Two_fields{
+            Two_fields{Float64, Float64},
+            Two_fields{Float64, Two_fields{Float64, Singleton{Float64}}},
+        },
+        Singleton{Float64},
+    )
+end
+
+@testset "fieldmatrix to scalar fieldmatrix unit tests" begin
+    FT = Float64
+    center_space, face_space = test_spaces(FT)
+    surface_space = Spaces.level(face_space, half)
+    seed!(1)
+    λ = 10
+    ᶜᶜmat1 = random_field(DiagonalMatrixRow{FT}, center_space) ./ λ .+ (I,)
+    ᶜᶠmat2 = random_field(BidiagonalMatrixRow{FT}, center_space) ./ λ
+    ᶠᶜmat2 = random_field(BidiagonalMatrixRow{FT}, face_space) ./ λ
+    ᶜᶜmat3 = random_field(TridiagonalMatrixRow{FT}, center_space) ./ λ .+ (I,)
+    ᶠᶠmat3 = random_field(TridiagonalMatrixRow{FT}, face_space) ./ λ .+ (I,)
+
+    e¹² = Geometry.Covariant12Vector(1, 2)
+    e³ = Geometry.Covariant3Vector(1)
+    e₃ = Geometry.Contravariant3Vector(1)
+
+    ρχ_unit = (; ρq_tot = 1, ρq_liq = 1, ρq_ice = 1, ρq_rai = 1, ρq_sno = 1)
+    ρaχ_unit =
+        (; ρaq_tot = 1, ρaq_liq = 1, ρaq_ice = 1, ρaq_rai = 1, ρaq_sno = 1)
+
+
+    ᶠᶜmat2_u₃_scalar = ᶠᶜmat2 .* (e³,)
+    ᶜᶠmat2_scalar_u₃ = ᶜᶠmat2 .* (e₃',)
+    ᶠᶠmat3_u₃_u₃ = ᶠᶠmat3 .* (e³ * e₃',)
+    ᶜᶠmat2_ρχ_u₃ = map(Base.Fix1(map, Base.Fix2(⊠, ρχ_unit ⊠ e₃')), ᶜᶠmat2)
+    A = dycore_prognostic_EDMF_FieldMatrix(;
+        ᶜᶜmat1,
+        ᶜᶠmat2,
+        ᶠᶜmat2,
+        ᶜᶜmat3,
+        ᶠᶠmat3,
+        e¹²,
+        e³,
+        e₃,
+        ρχ_unit,
+        ρaχ_unit,
+        ᶜᶠmat2_ρχ_u₃,
+        ᶠᶠmat3_u₃_u₃,
+        ᶜᶠmat2_scalar_u₃,
+        ᶠᶜmat2_u₃_scalar,
+    )
+
+    scalar_keys = MatrixFields.get_scalar_keys(A)
+    @test length(scalar_keys) > length(keys(A))
+    @test all(
+        f -> f isa MatrixFields.UniformScaling || eltype(eltype(f)) <: FT,
+        getindex.(Ref(A), scalar_keys),
+    )
+    @test (@allocated MatrixFields.get_scalar_keys(A)) == 0
+    @test_opt MatrixFields.get_scalar_keys(A)
+    foreach(scalar_keys) do key
+        @test_opt A[key]
+        @test (@allocated A[key]) == 0
+    end
+
+    A_scalar = MatrixFields.scalar_fieldmatrix(A)
+    @test all(
+        f -> f isa MatrixFields.UniformScaling || eltype(eltype(f)) <: FT,
+        getindex.(Ref(A_scalar), scalar_keys),
+    )
+    @test all(scalar_keys) do key
+        entry = A_scalar[key]
+        entry isa MatrixFields.UniformScaling && return true
+        key′, entry′ =
+            filter(pair -> MatrixFields.is_child_value(key, pair[1]), pairs(A))[1]
+        if key != key′
+            parent(entry) isa SubArray ||
+                parent(entry) === parent(entry′) ||
+                return false
+        end
+        parent(parent(entry)) === parent(entry′) || return false
+        return true
+    end
```

Comment:
```suggestion
    for (key, entry) in MatrixFields.scalar_fieldmatrix(A)
        @test_all A[key] === entry
    end
```
This looks like it's testing a bunch of internal implementation details, but all we really need to check is that the parent arrays are identical (which we can do by checking for physical equality of the entries themselves). Also, I'd recommend writing loops over separate tests instead of combining the loops into `@test all(...)`, since the short-circuiting logic is a bit hard to read.